src/bind.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Bind charm business logic."""
+
+import logging
+
+from charms.operator_libs_linux.v2 import snap
+
+import constants
+import exceptions
+
+logger = logging.getLogger(__name__)
+
+
+class ReloadError(exceptions.BlockableError):
+    """Exception raised when unable to reload the service."""
+
+
+class StartError(exceptions.BlockableError):
+    """Exception raised when unable to start the service."""
+
+
+class StopError(exceptions.BlockableError):
+    """Exception raised when unable to stop the service."""
+
+
+class InstallError(exceptions.BlockableError):
+    """Exception raised when unable to install dependencies for the service."""
```

Comment:
What is the operator expected to do when these errors occur and the charm goes into BlockedState? Afaik BlockedState should be used when there is an obvious action (preferably using juju like relating  or applying configuration) to unblock the state. I don't think it should be used when the error is unexpected and the operator has to dig into troubleshooting.
See also https://discourse.charmhub.io/t/its-probably-ok-for-a-unit-to-go-into-error-state/13022/30 .Perhaps @gregory-schiano will be able to give his opinion here.

Reply:
I think they are handled in `charm.py` lines 39-70 and the charm goes in BlockedState.